### PR TITLE
📖  Add clusterclass contract changes to v1alpha3-v1alpha4 conversion guide

### LINKED ISSUE #5282: Update v1alpha3->alpha4 conversion guide with recent contract revisions
**Detailed Description**
This task involves updating the CAPI book's v1alpha3-v1alpha4 conversion guide to include recent changes in the CAPI contract.

As per https://github.com/kubernetes-sigs/cluster-api/pull/5264#issuecomment-923992674 changes made over the course of a release - e.g. clusterClass motivated changes to the CAPI contract made during v1alpha4 - should be included in the correct conversion documentation in the CAPI book where they impact providers.

 Two specific changes, 
>  #4949 on v1alpha3 -> v1alpha4 doc + the need of implementing InfrastructureClusterTemplate or ControlPlaneTemplate for ClusterClass **support**

are highlighted here, but there may be more to document on reviewing the CAPI changes over the full lifetime of v1apha4.

https://cluster-api.sigs.k8s.io/developer/providers/v1alpha3-to-v1alpha4.html

### PR DESCRIPTION
**Pull request by killianmuldoon**
Update docs to include some changes made to the ControlPlane contract to support ClusterClass.

Signed-off-by: killianmuldoon <kmuldoon@vmware.com>

Fixes #5282

### CODE REVIEW

**Review by killianmuldoon — COMMENTED** [review]
I think this should be cherry picked back to the 0.4 docs too.

File: docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md [hunk 1]
```diff
@@ -334,3 +334,22 @@ should be executed before this changes.
 - `spec.infrastructureTemplate` has been moved to `spec.machineTemplate.infrastructureRef`. Thus, cluster templates which include `KubeadmControlPlane`
 have to be adjusted accordingly.
 - `spec.nodeDrainTimeout` has been moved to `spec.machineTemplate.nodeDrainTimeout`.
+
+## Required spec and status fields for Control Planes using 'version' for ClusterClass support
+ControlPlane implementations using version must now [include a 'version' field as defined below in both its spec and its status](../architecture/controllers/control-plane.md#required-spec-fields-for-implementations-using-version).
+
+ `version` - is a string representing the Kubernetes version to be used
```
Comment (sbueringer) [inline docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:341]: Maybe it's better to drop the duplicate definition here as `.status.version` is defined differently

Comment (killianmuldoon) [reply docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md]: Thanks for pointing that out! I've updated with a separate definition for each field which should be useful to someone consuming these docs I think.

File: docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md [hunk 2]
```diff
@@ -334,3 +334,24 @@ should be executed before this changes.
...
+ControlPlane implementations using version must now [include a 'version' field as defined below in both its spec and its status](../architecture/controllers/control-plane.md#required-spec-fields-for-implementations-using-version).
+
+ `spec.version` - is a string representing the Kubernetes version to be used
+  by the control plane machines. The value must be a valid semantic version;
+  also if the value provided by the user does not start with the v prefix, it
+  must be added.
+
+`status.version` - is a string representing the minimum Kubernetes version for the control plane machines in the cluster.
+
+These fields are a requirement for ClusterClass compatibility.
```
Comment (fabriziopandini) [inline docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:348]:
```suggestion
Please note that implementing those field is a requirement for a control plane provider to be used with ClusterClass and managed topologies.
```

File: docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md [hunk 3]
```diff
@@ -334,3 +334,24 @@ should be executed before this changes.
...
+
+## Required spec fields for Control Planes using 'Machines' for ClusterClass support
+ControlPlane implementations that use an underlying MachineInfrastructure must now [include a 'machineTemplate' as defined below, with subordinate fields, in its Spec.](../architecture/controllers/control-plane.md#required-spec-fields-for-implementations-using-machines)
+
+
+ `machineTemplate` - is a struct containing details of the control plane
+  machine template.
+
+This field is a requirement for ClusterClass compatibility.
```
Comment (fabriziopandini) [inline docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:357]:
```suggestion
Please note that implementing those field for control plane providers using machines is a requirement for ClusterClass and managed topologies to work consistently across all the providers.
```

**Review by fabriziopandini — COMMENTED** [review]
Thanks for this effort!
Overall lgtm, only two nits to provide some more context about/when those changes are required

**Comment by killianmuldoon** [comment]
@fabriziopandini @sbueringer updated now.

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**